tools/handshake-runner/Simulation.cpp

```diff
@@ -106,22 +106,53 @@ Any readValueWithType(mlir::Type type, std::string in) {
     arg >> x;
     APFloat aparg(x);
     return aparg;
+  } else if (auto tupleType = type.dyn_cast<TupleType>()) {
+    char tmp;
+    arg >> tmp;
+    assert(tmp == '(' && "tuple should start with '('");
+    std::vector<Any> values;
+    unsigned size = tupleType.getTypes().size();
+    values.reserve(size);
+    // Parse element by element
+    for (unsigned i = 0; i < size; ++i) {
+      values.push_back(readValueWithType(tupleType.getType(i), arg));
+      // Consumes either the ',' or the ')'
+      arg >> tmp;
+    }
+    assert(tmp == ')' && "tuple should end with ')'");
+    return values;
   } else {
     assert(false && "unknown argument type!");
     return {};
   }
 }
 
-std::string printAnyValueWithType(mlir::Type type, Any &value) {
-  std::stringstream out;
+Any readValueWithType(mlir::Type type, std::string in) {
+  std::stringstream stream(in);
+  return readValueWithType(type, stream);
+}
+
+void printAnyValueWithType(llvm::raw_ostream &out, mlir::Type type,
+                           Any &value) {
   if (type.isa<mlir::IntegerType>() || type.isa<mlir::IndexType>()) {
     out << any_cast<APInt>(value).getSExtValue();
-    return out.str();
   } else if (type.isa<mlir::FloatType>()) {
     out << any_cast<APFloat>(value).convertToDouble();
-    return out.str();
   } else if (type.isa<mlir::NoneType>()) {
-    return "none";
+    out << "none";
+  } else if (auto tupleType = type.dyn_cast<mlir::TupleType>()) {
+    auto values = any_cast<std::vector<llvm::Any>>(value);
+    out << "(";
+    bool fst = true;
+    for (auto [type, value] : llvm::zip(tupleType.getTypes(), values)) {
+      if (fst)
+        fst = false;
+      else
+        out << ", ";
+      printAnyValueWithType(out, type, value);
+    }
```

Comment:
`llvm::interleaveComma` is your friend here.

Reply:
Didn't know about this. Thanks for pointing this out to me.